src/Planner/PlannerJoins.cpp

```diff
@@ -357,24 +349,58 @@ void buildJoinClause(
                 join_clause.addCondition(left_expression_side, node);
             }
         }
+        else
+        {
+            auto support_mixed_join_condition = planner_context->getQueryContext()->getSettingsRef().enable_mixed_join_condition;
```

Comment:
Is the only purpose of the setting is to throw exception? In that case we can just allow new syntax. Or introduce it as experimental for some time, `allow_experimental_join_condition`